### ISSUE TITLE
Consider offering a full jitter backoff policy

### ISSUE THREAD

**Issue by dbolduc** [googleapis/google-cloud-cpp#8755]
Here is a write up on [full jitter](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/). Apparently the java clients [do this](https://github.com/googleapis/gax-java/blob/main/gax/src/main/java/com/google/api/gax/retrying/ExponentialRetryAlgorithm.java#L253-L256).

With pure exponential backoff policy we might have backoff ranges like:
req1: (.5, 1)
req2: (1, 2)
req3: (2, 4)
....

With full jitter, these backoff ranges would look like:
req1: (0, 1)
req2: (0, 2)
req3: (0, 4)
....

(make up your own units)

**Comment by coryan** [comment]
For some services (storage comes to mind) a minimum back off is recommended.

**Comment by coryan** [comment]
@dbolduc is going to research some more and make a recommendation.

**Comment by dbolduc** [comment]
This is the code to run the simulation in the article: https://github.com/aws-samples/aws-arch-backoff-simulator/blob/master/src/backoff_simulator.py

I added our backoff algorithm:
```diff
34a35,39
> class ExpoBackoffCloudCpp(Backoff):
>     def backoff(self, n):
>         v = self.expo(n)
>         return random.uniform(v/2, v)
> 
```

And a MinJitter (i.e. full jitter, but it doesn't start at 0)
```diff
39a45,49
> class ExpoBackoffMinJitter(Backoff):
>     def backoff(self, n):
>         v = self.expo(n)
>         return random.uniform(self.base, v)
> 
```

Results:
![image](https://user-images.githubusercontent.com/23088558/189767201-5b1d98fb-b4d1-47ee-b3fa-ce8d183c68a1.png)

![image](https://user-images.githubusercontent.com/23088558/189767255-5261e03c-dc60-4336-b0d1-42d2f0e9b38b.png)

To summarize (with these exact settings, in this exact model), min jitter is indistinguishable from full jitter. Our strategy is indistinguishable from the jitter strategies in terms of total calls, but takes more time to complete the work.

I will phone a friend before making a recommendation...

**Comment by dbolduc** [comment]
Seems like other client library languages do full jitter only. Given the supposed, slight performance benefit, I think we should implement min-jitter (which is like full jitter, but slightly more general).

I think we should modify the implementation of the existing `ExponentialBackoffPolicy`, with minimal behavior changes for the current constructor. I think this means: the upper bound of the range should match, and the initial lower bound should match. (The lower bound will not match for subsequent calls to `OnCompletion()`. This is a change of behavior, but not a breaking change of our API.).

I think our default backoff policy should use full jitter.

## Background

Our API accepts three parameters:
- minimum delay
- maximum delay (Note that this is an overall maximum.)
- scaling factor

https://github.com/googleapis/google-cloud-cpp/blob/9737c4bfc27eea6c593dcbe2d59af41d266acc0c/google/cloud/internal/backoff_policy.h#L125-L128

Aside: I don't agree with the range it sets. I think we should multiply by `scaling`, not `2`. Whatever. https://github.com/googleapis/google-cloud-cpp/blob/9737c4bfc27eea6c593dcbe2d59af41d266acc0c/google/cloud/internal/backoff_policy.h#L131

Min-jitter requires four parameters:
- minimum delay (0 in the case of full jitter)
- initial delay upper bound
- maximum delay
- scaling factor

## Design / Work:

I would send one PR to do each of the following:

### Implement min-jitter
  - add a new constructor + a `minimum_delay_` member.
    - I'd probably rename  `s/current_range_delay_/current_range_upper_bound_/`, too.
  - map the parameters for the existing ctor to the members. (requires thinking!)
  - update `clone()` and `OnCompletion()` implementations.
  - add/update [tests](https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/internal/backoff_policy_test.cc)
  - update [documentation](https://github.com/googleapis/google-cloud-cpp/blob/9737c4bfc27eea6c593dcbe2d59af41d266acc0c/google/cloud/internal/backoff_policy.h#L89-L124)

### Update library defaults

- tell the generator to use full-jitter:
https://github.com/googleapis/google-cloud-cpp/blob/main/generator/internal/option_defaults_generator.cc#L156-L157
- run the generator

**Comment by coryan** [comment]
Can we close this?

**Comment by alevenberg** [comment]
Yes
